### PR TITLE
Prototype of Response<T> and T clients

### PR DESCRIPTION
**Pull request by alzimmermsft**
Prototype for #3696 

ConfigurationClient and ConfigurationAsyncClient now return T instead of Response<T> and ConfigurationRawClient was added to support returning Response<T>

### CODE REVIEW

**Comment by JonathanGiles** [comment]
How does this API feel to you compared to what existed before?

**Comment by alzimmermsft** [comment]
Removing the need to extract the value from the Response is nicer. The change I made for the builder I have mixed feelings on but I also don't like the idea of three nearly identical classes that do the same thing.

I've been playing with it more and ran into some issues when the value of the response is null, using map on a null causes a crash, using flatMap with Mono.justOrEmpty is also bad since that returns as an onComplete status which causes issues for testing.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 1]
```diff
@@ -37,41 +25,31 @@
...
-public final class ConfigurationAsyncClient extends ServiceClient {
-    private final ServiceLogger logger = new ServiceLogger(ConfigurationAsyncClient.class);
-
-    private static final String ETAG_ANY = "*";
-    private static final String RANGE_QUERY = "items=%s";
-
-    private final String serviceEndpoint;
-    private final ConfigurationService service;
+public final class ConfigurationAsyncClient {
+    private final ConfigurationRawClient client;
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: Let's sync up on naming - not convinced 'Raw' is right, but let's see what else pops up.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 2]
```diff
@@ -123,17 +101,9 @@ public static ConfigurationAsyncClientBuilder builder() {
...
-        // This service method call is similar to setSetting except we're passing If-Not-Match = "*". If the service
-        // finds any existing configuration settings, then its e-tag will match and the service will return an error.
-        return service.setKey(serviceEndpoint, setting.key(), setting.label(), setting, null, getETagValue(ETAG_ANY))
-            .doOnRequest(ignoredValue -> logger.asInformational().log("Adding ConfigurationSetting - {}", setting))
-            .doOnSuccess(response -> logger.asInformational().log("Added ConfigurationSetting - {}", response.value()))
-            .onErrorMap(ConfigurationAsyncClient::addSettingExceptionMapper)
-            .doOnError(error -> logger.asWarning().log("Failed to add ConfigurationSetting - {}", setting, error));
+    public Mono<ConfigurationSetting> addSetting(ConfigurationSetting setting) {
+        return client.addSetting(setting)
+            .flatMap(response -> Mono.justOrEmpty(response.value()));
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: If we take this approach this code pattern will become so common that we will want to optimise it. I propose you create some static utility method named `unwrapResponse` or something that would be called along the lines of `return unwrapResponse(client.addSetting(setting))`

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java [hunk 3]
```diff
@@ -39,109 +55,176 @@
...
      *
      * <p>
      * If {@link ConfigurationClientBuilder#pipeline(HttpPipeline) pipeline} is set, then the {@code pipeline} and
      * {@link ConfigurationClientBuilder#serviceEndpoint(String) serviceEndpoint} are used to create the
-     * {@link ConfigurationClient client}. All other builder settings are ignored.</p>
+     * {@link ConfigurationRawClient client}. All other builder settings are ignored.
+     * </p>
      *
-     * @return A ConfigurationClient with the options set from the builder.
+     * @return A ConfigurationAsyncClient with the options set from the builder.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: This should say ConfigurationRawClient I believe.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationRawClient.java [hunk 4]
```diff
@@ -0,0 +1,552 @@
...
+ * in Azure App Configuration Store. Operations allowed by the client are adding, retrieving, updating, and deleting
+ * ConfigurationSettings, and listing settings or revision of a setting based on a {@link SettingSelector filter}.
+ *
+ * <p><strong>Instantiating an Asynchronous Configuration Client</strong></p>
+ *
+ * <pre>
+ * ConfigurationAsyncClient client = ConfigurationAsyncClient.builder()
+ *     .credentials(new ConfigurationClientCredentials(connectionString))
+ *     .build();
+ * </pre>
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationRawClient.java]: We will need to revisit all the reference docs if this change is made.

Comment (alzimmermsft) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationRawClient.java]: Definitely will need something to make maintaining the examples easier.

**Comment by alzimmermsft** [comment]
Closing PR as we are no longer pursuing this route.

**Pull request closed without merging**